pkg/admission/apiresourceschema/admission_test.go

```diff
@@ -0,0 +1,163 @@
+		apisv1alpha1.Resource("apiresourceschemas").WithVersion("v1alpha1"),
+		"",
+		admission.Create,
+		&metav1.CreateOptions{},
+		false,
+		&user.DefaultInfo{},
+	)
+}
+
+// nolint:deadcode,unused
+func updateAttr(s, old *apisv1alpha1.APIResourceSchema) admission.Attributes {
+	return admission.NewAttributesRecord(
+		s,
+		old,
+		tenancyv1alpha1.Kind("APIResourceSchema").WithVersion("v1alpha1"),
+		"",
+		s.Name,
+		tenancyv1alpha1.Resource("apiresourceschemas").WithVersion("v1alpha1"),
+		"",
+		admission.Update,
+		&metav1.CreateOptions{},
+		false,
+		&user.DefaultInfo{},
+	)
+}
+
+func TestValidate(t *testing.T) {
+	tests := []struct {
+		name           string
+		attr           admission.Attributes
+		expectedErrors []string
+	}{
+		{
+			name: "an APIResourceSchema can pass admission",
+			attr: createAttr(unmarshalOrDie(`
+apiVersion: apis.kcp.sh/v1alpha1
+kind: APIResourceSchema
+metadata:
+  name: july.cowboys.wild.west
+spec:
+  group: wild.west
+  names:
+    plural: cowboys
+    singular: cowboy
+    kind: Cowboy
+    listKind: CowboyList
+  scope: Cluster
+  versions:
+  - name: v1
+    served: true
+    storage: true
+    subresources:
+      status: {}
+    schema:
+      type: object
+      properties:
+        spec:
+          type: object
+            `)),
+		},
+		{
+			name: "an APIResourceSchema can fail admission",
+			attr: createAttr(unmarshalOrDie(`
+apiVersion: apis.kcp.sh/v1alpha1
+kind: APIResourceSchema
+metadata:
+  name: may.2022.cowboys.wild.west
+spec:
+  group: wild.west
+  scope: Clustered
+  names:
+	plural: cowboys
+  versions:
+  - name: v1-rc17
+    served: true
+    storage: false
+  - name: v2
+    served: true
+    storage: true
+	schema:
+	  type: thing
+            `)),
+			expectedErrors: []string{
+				"metadata.name: Invalid value: \"may.2022.cowboys.wild.west\": must match patch ^[a-z]([-a-z0-9]*[a-z0-9])? in front of .<resource>.<group>",
+				"spec.versions[0].schema: Required value: schemas are required",
+				"spec.versions[1].schema.openAPIV3Schema.type: Unsupported value: \"thing\": supported values: \"array\", \"boolean\", \"integer\", \"number\", \"object\", \"string\"",
+				"spec.versions[1].schema.openAPIV3Schema.type: Invalid value: \"thing\": must be object at the root",
+				"spec.names.singular: Required value",
+				"spec.names.kind: Required value",
+				"spec.names.listKind: Required value",
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			o := &apiResourceSchemaValidation{
+				Handler: admission.NewHandler(admission.Create, admission.Update),
+			}
+			ctx := request.WithCluster(context.Background(), request.Cluster{Name: "root:org"})
+			err := o.Validate(ctx, tt.attr, nil)
+			wantErr := len(tt.expectedErrors) > 0
+			if (err != nil) != wantErr {
+				t.Fatalf("Validate() error = %v, wantErr %v", err, wantErr)
+			}
+			if err != nil {
+				t.Logf("Got admission errors: %v", err)
+				for _, expected := range tt.expectedErrors {
```

Comment:
Would validation ever return these in a different order?

Reply:
technically not, but if it does no need to fail the test.